cmd/completion/main.go

```diff
@@ -31,46 +42,96 @@ func init() {
+	flag.Var(&cosignRepositories, "cosign-repositories", "Cosign signing repository of the form 'secretname=registry.example.com/project'")
+	flag.Var(&cosignDockerMediaTypes, "cosign-docker-media-types", "Cosign signing with legacy docker media types of the form 'secretname=1'")
 	logger = log.New(os.Stdout, "", 0)
 }
 
 func main() {
 	flag.Parse()
 
-	if notaryV1URL != "" {
-		creds, err := dockercreds.ParseMountedAnnotatedSecrets(registrySecretsDir, dockerCredentials)
+	var report lifecycle.ExportReport
+	_, err := toml.DecodeFile(reportFilePath, &report)
+	if err != nil {
+		logger.Fatalf("toml decode: %v", err)
+	}
+
+	creds, err := dockercreds.ParseMountedAnnotatedSecrets(registrySecretsDir, dockerCredentials)
+	if err != nil {
+		logger.Fatal(err)
+	}
+
+	for _, c := range append(dockerCfgCredentials, dockerConfigCredentials...) {
+		credPath := filepath.Join(registrySecretsDir, c)
+
+		dockerCfgCreds, err := dockercreds.ParseDockerPullSecrets(credPath)
 		if err != nil {
 			logger.Fatal(err)
 		}
 
-		for _, c := range append(dockerCfgCredentials, dockerConfigCredentials...) {
-			credPath := filepath.Join(registrySecretsDir, c)
+		for domain := range dockerCfgCreds {
+			logger.Printf("Loading secret for %q from secret %q at location %q", domain, c, credPath)
+		}
 
-			dockerCfgCreds, err := dockercreds.ParseDockerPullSecrets(credPath)
-			if err != nil {
-				logger.Fatal(err)
-			}
+		creds, err = creds.Append(dockerCfgCreds)
+		if err != nil {
+			logger.Fatal(err)
+		}
 
-			for domain := range dockerCfgCreds {
-				logger.Printf("Loading secret for %q from secret %q at location %q", domain, c, credPath)
-			}
+		homeDir, err := os.UserHomeDir()
+		if err != nil {
+			logger.Fatalf("error obtaining home directory: %v", err)
+		}
 
-			creds, err = creds.Append(dockerCfgCreds)
-			if err != nil {
-				logger.Fatal(err)
-			}
+		err = creds.Save(path.Join(homeDir, ".docker", "config.json"))
+		if err != nil {
+			logger.Fatalf("error writing docker creds: %v", err)
 		}
+	}
 
+	cosignSigner := cosigner.ImageSigner{
+		Logger: logger,
+	}
+
+	annotations := mapKeyValueArgs(cosignAnnotations)
+	cosignRepositoryOverrides := mapKeyValueArgs(cosignRepositories)
+	cosignDockerMediaTypesOverrides := mapKeyValueArgs(cosignDockerMediaTypes)
+
+	ctx := context.Background()
+	if err := cosignSigner.Sign(ctx, report, annotations, cosignRepositoryOverrides, cosignDockerMediaTypesOverrides); err != nil {
+		logger.Fatalf("cosignSigner sign: %v\n", err)
+	}
+
+	if notaryV1URL != "" {
 		signer := notary.ImageSigner{
 			Logger:  logger,
 			Client:  &registry.Client{},
 			Factory: &notary.RemoteRepositoryFactory{},
 		}
-		if err := signer.Sign(notaryV1URL, notarySecretDir, reportFilePath, creds); err != nil {
+		if err := signer.Sign(notaryV1URL, notarySecretDir, report, creds); err != nil {
 			logger.Fatal(err)
 		}
 	}
 
 	logger.Println("Build successful")
 }
+
+func mapKeyValueArgs(args flaghelpers.CredentialsFlags) map[string]interface{} {
```

Comment:
can this return `map[string]string` ?

Reply:
cosign's `cli.signCmd` takes in `map[string]interface{}` for it's annotation parameter. I didn't want two different functions, one returning `map[string]interface{}` and one returning `map[string]string` or having to convert between those two map types. It was easier to just do a `fmt.Sprintf("%s", <variable>)` cast when the `cosignRepository` and `cosignDockerMediaTypes` maps were used.

Would you recommend I still do a `map[string]string` and do a loop copy to a `map[string]interface{}` for annotations later?
If the current casting option is fine, is there a better way to cast an `interface{}` to a string? I was also looking at `variable.(string)` instead of using `fmt`.

Reply:
I see, I think that's a good enough reason to keep it as `map[string]interface{}`

Reply:
`fmt` works fine to cast

Reply:
resolving